Dockerfile

```diff
@@ -38,21 +35,43 @@ RUN chmod g=u /var/www
 # Set working directory
 WORKDIR /var/www
 
-# Change to composer user
+# Copy files needed by composer
+COPY ./composer.json /var/www/composer.json
+COPY ./composer.lock /var/www/composer.lock
+
+# Copy only files required by composer post-autoload-dump script (artisan package:discover --ansi)
+COPY ./app/Console/Kernel.php /var/www/app/Console/Kernel.php
+COPY ./app/Exceptions/Handler.php /var/www/app/Exceptions/Handler.php
+COPY ./app/Providers /var/www/app/Providers
+COPY ./bootstrap /var/www/bootstrap
+COPY ./config /var/www/config
+COPY ./database /var/www/database
+COPY ./routes /var/www/routes
+COPY ./storage /var/www/storage
+COPY ./artisan /var/www/artisan
+
+# Update permissions of copied files
+RUN chown -R composer:www-data /var/www
+RUN chmod -R g=u /var/www
+
+# Run Composer Install as user composer
 USER composer
-
-# Run Composer Install
 RUN /usr/local/bin/composer install
 
-# Change current user to www
+# As long as you haven't changed any of the above files since last build, image should be cached beyond this point
+# Yay, no more unnecessary composer installs!!!
+
+# Change current user to www-data
 USER www-data
 
-# Since .env is in .dockerignore, we have to get it from somewhere
-RUN cp .env.example .env
+# artisan key:generate requires .env, but it's in .dockerignore so get it from here
+COPY --chown=www-data:www-data ./.env.example /var/www/.env
 
 # Generate App Key
```

Comment:
We may want to consider moving this into the init script - depends if you want the key in the image.